### ISSUE TITLE
Exploring `immutable` keyword to save gas

### ISSUE THREAD

**Issue by ChihChengLiang** [thehubbleproject/hubble-contracts#455]
### What's wrong

Since we already upgrade to Solidity 0.6, there's a nice `immutable` keyword we can consider using.

Here's how it works. We usually have variables whose values won't be changed once the contract is deployed. However, their values must be determined at the deployment stage. We can't hardcode those values in the codebase, so we can't use the `constant` keyword for them.

In our use case, deployment parameters and some contract addresses that a contract depends on are such cases.

The immutable keyword replaces the variable references in the runtime bytecode with the value defined in the constructor. No need to SLOAD the variable when querying its value.

Example:
before
```
contract Rollup {
    uint256 public paramStakeAmount;

    constructor(uint256 stakeAmount){
        paramStakeAmount = stakeAmount;
    }
    function rollback() internal {
        ...
        // SLOAD paramStakeAmount for value
        uint256 slashedAmount = nActual.mul(paramStakeAmount);
        ...
    }
}
```
after
```
contract Rollup {
    uint256 public immutable paramStakeAmount;

    // deployed with stakeAmount = 10
    constructor(uint256 stakeAmount){
        paramStakeAmount = stakeAmount;
    }
    function rollback() internal {
        ...
        // paramStakeAmount is replaced with 10 in the runtime bytecode. No SLOAD
        uint256 slashedAmount = nActual.mul(paramStakeAmount);
        ...
    }
}
```

### Benefits

This saves us one SLOAD for each potential immutable referenced in a function.

### Concerns

Add a complication to reason the contract bytecode. We need to think about the side effects.

**Comment by ChihChengLiang** [comment]
https://github.com/thehubbleproject/hubble-contracts/blob/4ff3955f3faea7e61593d05a24fb8fd70ca7095f/contracts/rollup/Rollup.sol#L414
> No immutable variables are used in the Rollup contract. We propose using immutables to reduce a number of SLOADs.

**Comment by ChihChengLiang** [comment]
Closed by #491 and #497
